server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeRole.java

```diff
@@ -242,4 +261,84 @@ public void validateRoles(final List<DiscoveryNodeRole> roles) {
 
     }
 
+    // the set of possible roles
+    private static final SortedSet<DiscoveryNodeRole> ROLES;
+
+    /**
+     * The possible node roles.
+     *
+     * @return an ordered, immutable set of possible roles
+     */
+    public static SortedSet<DiscoveryNodeRole> roles() {
+        return ROLES;
+    }
+
+    // a map from role names to their role representations
+    private static final Map<String, DiscoveryNodeRole> ROLE_MAP;
+
+    static {
+        final List<Field> roleFields = Arrays.stream(DiscoveryNodeRole.class.getFields())
+            .filter(f -> f.getType().equals(DiscoveryNodeRole.class))
+            .collect(Collectors.toUnmodifiableList());
+        // this will detect duplicate role names
+        final Map<String, DiscoveryNodeRole> roleMap = roleFields.stream()
+            .map(f -> {
+                try {
+                    return (DiscoveryNodeRole) f.get(null);
+                } catch (final IllegalAccessException e) {
+                    throw new AssertionError(e);
+                }
+            })
+            .collect(Collectors.toMap(DiscoveryNodeRole::roleName, Function.identity()));
+        assert roleMap.size() == roleFields.size() :
+            "roles by name [" + roleMap + "], role fields [" + roleFields + "]";
+        final SortedSet<DiscoveryNodeRole> roles = roleMap.values().stream().collect(Sets.toUnmodifiableSortedSet());
+        // collect the abbreviation names into a map to ensure that there are not any duplicate abbreviations
+        final Map<String, DiscoveryNodeRole> abbreviations = roles
+            .stream()
+            .collect(Collectors.toUnmodifiableMap(DiscoveryNodeRole::roleNameAbbreviation, Function.identity()));
+        assert roleMap.size() == abbreviations.size() :
+            "roles by name [" + roleMap + "], roles by name abbreviation [" + abbreviations + "]";
```

Comment:
Same comment here about duplicates and assertions

Reply:
Same reply about the other assertion.